src/Storages/NamedCollections.cpp

```diff
@@ -0,0 +1,545 @@
+
+        if (to_config.has(to_path))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Key `{}` already exists", to_path);
+
+        if constexpr (std::is_same_v<T, String>)
+            to_config.setString(to_path, from_config.getString(from_path));
+        else if constexpr (std::is_same_v<T, std::string>)
+            to_config.setString(to_path, from_config.getString(from_path));
+        else if constexpr (std::is_same_v<T, UInt64>)
+            to_config.setUInt64(to_path, from_config.getUInt64(from_path));
+        else if constexpr (std::is_same_v<T, Int64>)
+            to_config.setInt64(to_path, from_config.getInt64(from_path));
+        else if constexpr (std::is_same_v<T, Float64>)
+            to_config.setDouble(to_path, from_config.getDouble(from_path));
+        else
+            throw Exception(
+                ErrorCodes::NOT_IMPLEMENTED,
+                "Unsupported type in copyConfigValue(). "
+                "Supported types are String, UInt64, Int64, Float64");
+    }
+
+    static void removeConfigValue(
+        Poco::Util::AbstractConfiguration & config,
+        const std::string & path)
+    {
+        if (!config.has(path))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "No such key `{}`", path);
+        config.remove(path);
+    }
+
+    static ConfigurationPtr createEmptyConfiguration(const std::string & root_name)
+    {
+        using DocumentPtr = Poco::AutoPtr<Poco::XML::Document>;
+        DocumentPtr xml_document(new Poco::XML::Document());
+        xml_document->appendChild(xml_document->createElement(root_name));
+        ConfigurationPtr config(new Poco::Util::XMLConfiguration(xml_document));
+        return config;
+    }
+};
+
+NamedCollection::NamedCollection(
+    const Poco::Util::AbstractConfiguration & config,
+    const std::string & collection_path,
+    const Keys & keys)
+    : NamedCollection(std::make_unique<Impl>(config, collection_path, keys))
+{
+}
+
+NamedCollection::NamedCollection(ImplPtr pimpl_)
+    : pimpl(std::move(pimpl_))
+{
+}
+
+NamedCollectionPtr NamedCollection::create(
+    const Poco::Util::AbstractConfiguration & config,
+    const std::string & collection_name)
+{
+    const auto collection_prefix = getCollectionPrefix(collection_name);
+    std::queue<std::string> enumerate_input;
+    std::set<std::string> enumerate_result;
+
+    enumerate_input.push(collection_prefix);
+    collectKeys(config, enumerate_input, enumerate_result);
+
+    /// Collection does not have any keys.
+    /// (`enumerate_result` == <collection_path>).
+    const bool collection_is_empty = enumerate_result.size() == 1;
+    std::set<std::string> keys;
+    if (!collection_is_empty)
+    {
+        /// Skip collection prefix and add +1 to avoid '.' in the beginning.
+        for (const auto & path : enumerate_result)
+            keys.emplace(path.substr(std::strlen(collection_prefix.data()) + 1));
```

Comment:
bit confused, why not use `collection_prefix.size()`? 😮

Reply:
Oops :sweat_smile: